watertap/core/plugins/tests/test_solvers.py

```diff
@@ -0,0 +1,148 @@
+#
+###############################################################################
+
+import pytest
+import pyomo.environ as pyo
+
+from pyomo.common.errors import ApplicationError
+from idaes.core.util.scaling import (set_scaling_factor, get_scaling_factor,
+        constraints_with_scale_factor_generator, unscaled_constraints_generator)
+from idaes.core.util import get_solver
+from watertap.core.plugins.solvers import IpoptWaterTAP
+
+class TestIpoptWaterTAP:
+    @pytest.fixture(scope="class")
+    def m(self):
+        m = pyo.ConcreteModel()
+        m.a = pyo.Var(initialize=1)
+        m.b = b = pyo.Block()
+        b.a = pyo.Var([1,2])
+
+        m.c = pyo.Constraint(expr=(0, 1./(m.a**2) ,1))
+        b.c = pyo.Constraint([1,2], rule= lambda b,i: (i-1, b.a[i], i+2))
+        b.d = pyo.Constraint(expr=b.a[1]**4 + b.a[2]**4 <= 4)
+        set_scaling_factor(b.d, 1e6)
+
+        b.o = pyo.Expression(expr=sum(b.a)**2)
+
+        m.o = pyo.Objective(expr=m.a+b.o)
+
+        return m
+
+    @pytest.fixture(scope="class")
+    def s(self):
+        return pyo.SolverFactory('ipopt-watertap')
+    
+    @pytest.mark.unit
+    def test_pyomo_registration(self, s):
+        assert s.__class__ is IpoptWaterTAP
+
+    @pytest.mark.unit
+    def test_idaes_registration(self):
+        assert get_solver().__class__ is IpoptWaterTAP
+
+    @pytest.mark.unit
+    def test_presolve_scales_constraints(self, m, s):
+        s._presolve(m, tee=True)
+        for c, sf in s._scaling_cache:
+            if c is m.b.d:
+                assert sf == 1e6
+            else:
+                assert sf is None
+        
+        cons_with_sf = [c for c,_ in constraints_with_scale_factor_generator(m)]
+        assert m.c in cons_with_sf
+        assert m.b.c[1] in cons_with_sf
+        assert m.b.c[2] in cons_with_sf
+        assert m.b.d in cons_with_sf
+
+        assert s._model is m
+
+    @pytest.mark.unit
+    def test_postsolve_unscaled_constraints_cleanup(self, m, s):
+        assert hasattr(s, '_postsolve')
+        # for the Pyomo implementation
+        try:
+            s._postsolve()
+        except AttributeError:
+            pass
+
+        assert not hasattr(s, '_scaling_cache')
+        
+        cons_with_sf = list(constraints_with_scale_factor_generator(m))
+        assert cons_with_sf == [(m.b.d, 1e6)]
+
+        assert not hasattr(s, '_model')
+
+    @pytest.mark.unit
+    def test_option_absorption(self, m, s):
+        s.options['ignore_variable_scaling'] = True
+        results = s.solve(m, tee=True)
+        pyo.assert_optimal_termination(results)
+        del s.options['ignore_variable_scaling']
+
+    @pytest.mark.unit
+    def test_get_option(self, s):
+        s.options['ignore_constraint_scaling'] = True
+
+        assert s._get_option('ignore_constraint_scaling', False) is True
+        assert 'ignore_constraint_scaling' not in s.options
+        assert s._get_option('ignore_constraint_scaling', False) is False
```

Comment:
Not clear what's happening here in the first assert. It looks like `ignore_constraint_scaling` was removed from options (from the pop?) but why assert `s._get_option('ignore_constraint_scaling', **False**)`? What would happen if you had `s._get_option('ignore_constraint_scaling', **True**)`? I am guessing that the last assert is to double-check that the option no longer exists although you did so with the second assert.

Reply:
The `_get_option` helper both pops and fetches the named option in the first argument, with the default specified as the second argument if the option doesn't exist. Since the `'ignore_constraint_scaling'` option is **set**, this helper should return the set value, as opposed to the default (`False`) (the first `assert`), and it should **pop** the options (the second `assert`). Now that the option no longer exists, the third assert verifies that `_get_option` returns the specified default when the option does not exist.

Reply:
Ah! Makes perfect sense now =)